packages/renderer/server/renderVideo.ts

```diff
@@ -0,0 +1,241 @@
+      configFile: resolvedConfigPath,
+      server: {
+        port: fixedPort,
+      },
+      plugins: [rendererPlugin(params)],
+    }).then(server => server.listen()),
+  ]);
+
+  if (!server.httpServer) {
+    throw new Error('HTTP server is not initialized');
+  }
+  const address = server.httpServer.address();
+  const port = address && typeof address === 'object' ? address.port : null;
+  if (port === null) {
+    throw new Error('Server address is null');
+  }
+
+  return {browser, server, port};
+}
+
+async function renderVideoOnPage(
+  browser: Browser,
+  server: ViteDevServer,
+  url: string,
+) {
+  const page = await browser.newPage();
+  if (!server.httpServer) {
+    throw new Error('HTTP server is not initialized');
+  }
+  const address = server.httpServer.address();
+  const port = address && typeof address === 'object' ? address.port : null;
+  if (port === null) {
+    throw new Error('Server address is null');
+  }
+
+  // Attach logs from puppeteer to the console
+  page.on('console', msg => {
+    for (let i = 0; i < msg.args().length; ++i) {
+      console.log(`${port}: ${msg.args()[i]}`);
+    }
+  });
+
+  page.exposeFunction('logProgress', (progress: number) => {
+    const percentage = Math.floor(progress * 100);
+    const barLength = 20;
+    const filledLength = Math.floor((percentage / 100) * barLength);
+    const bar = '█'.repeat(filledLength) + '-'.repeat(barLength - filledLength);
+    const consoleLine = process.stdout.rows + port - 9000;
+
+    readline.cursorTo(process.stdout, 0, consoleLine);
+    readline.clearLine(process.stdout, 0);
+    process.stdout.write(
+      `Progress for worker on port ${port}: [${bar}] ${percentage}%`,
+    );
+  });
+
+  const renderingComplete = new Promise<void>((resolve, reject) => {
+    page.exposeFunction('onRenderComplete', async () => {
+      await Promise.all([browser.close(), server.close()]);
+      resolve();
+    });
+
+    page.exposeFunction('onRenderFailed', async (errorMessage: string) => {
+      await Promise.all([browser.close(), server.close()]);
+      console.error('Rendering failed:', errorMessage);
+      reject(new Error(errorMessage));
+    });
+  });
+
+  await page.goto(url);
+
+  return renderingComplete;
+}
+
+async function cleanup(
+  projectName: string,
+  numOfWorkers: number,
+  hiddenFolderId: string,
+) {
+  const cleanupFolders = [];
+  const cleanupFiles = [];
+  for (let i = 0; i < numOfWorkers; i++) {
+    cleanupFolders.push(
+      `${os.tmpdir()}/revideo-${projectName}-${i}-${hiddenFolderId}`,
+    );
+    cleanupFiles.push(
+      path.join(process.cwd(), `output/${projectName}-${i}.mp4`),
+    );
+  }
+
+  cleanupFiles.push(
+    path.join(process.cwd(), `output/${projectName}-audio.wav`),
+  );
+  cleanupFiles.push(
+    path.join(process.cwd(), `output/${projectName}-visuals.mp4`),
+  );
+
+  for (const folder of cleanupFolders) {
+    try {
+      await fs.promises.rm(folder, {recursive: true, force: true});
```

Comment:
We could parallelize this with an array and Promise.all(). Feel free to ignore this though, I don't think we would notice any difference.